src/flash.rs

```diff
@@ -2,6 +2,242 @@
+            true => Ok(()),
+            false => Err(Error::LockError),
+        }
+    }
+
+    fn valid_address(&self, offset: u32) -> Result<()> {
+        if FLASH_START + offset > FLASH_END {
+            Err(Error::AddressLargedThanFlash)
+        }
+        else if offset % 2 != 0 {
+            Err(Error::AddressMisAligned)
+        }
+        else {
+            Ok(())
+        }
+    }
+
+    fn valid_length(&self, length: u32) -> Result<()> {
+        if length > self.sector_sz.size() as u32 {
+            Err(Error::LengthTooLong)
+        }
+        else if length % 2 != 0 {
+            Err(Error::LengthNotMultiple2)
+        }
+        else {
+            Ok(())
+        }
+    }
+
+    pub fn page_erase(&mut self, start_offset: u32) -> Result<()> {
+        self.valid_address(start_offset)?;
+
+        self.unlock()?;
+
+        // Set Page Erase
+        self.flash.cr.cr().modify(|_, w| w.per().set_bit() );
+
+        // Write address bits
+        // NOTE(unsafe) This sets the sector address in the Address Register
+        unsafe { self.flash.ar.ar().write(|w| w.far().bits(FLASH_START + start_offset) ); }
+
+        // Start Operation
+        self.flash.cr.cr().modify(|_, w| w.strt().set_bit() );
+
+        // Wait
+        while self.flash.sr.sr().read().bsy().bit_is_set() { }
+
+        // Check for errors
+        let sr = self.flash.sr.sr().read();
+
+        // Remove Page Erase Operation bit
+        self.flash.cr.cr().modify(|_, w| w.per().clear_bit() );
+
+        // Re-lock flash
+        self.lock()?;
+
+        for idx in 0..self.sector_sz.size() -1 {
+            let write_address = (FLASH_START + start_offset + idx as u32) as *mut u16;
+            let verify: u16 = unsafe { core::ptr::read_volatile(write_address) };
+            if verify != 0xFFFF {
+                return Err(Error::VerifyError);
+            }
+        }
+        if sr.wrprterr().bit_is_set() {
+            Err(Error::EraseError)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Erase Flash from FLASH_START + start_offset to length
+    pub fn erase(&mut self, start_offset: u32, length: usize) -> Result<()> {
+        if start_offset + length as u32 > self.flash_sz.size() {
+            return Err(Error::LengthTooLong);
+        }
+
+        for offset in (start_offset..start_offset + length as u32).step_by(self.sector_sz.size() as usize) {
+            self.page_erase(offset)?;
+        }
+
+        Ok(())
+    }
+
+    /// Retrieve an slice of at most 1 sector of data from FLASH_START +offset
+    pub fn read(&mut self, offset: u32, length: usize) -> Result<&'static [u8]> {
+        self.valid_address(offset)?;
+        self.valid_length(length as u32)?;
+
+        let address = (FLASH_START + offset) as *const _;
+
+        // NOTE(unsafe) grants read access to the FLASH
+        unsafe {
+            Ok(core::slice::from_raw_parts::<'static, u8>(address, length))
```

Comment:
Is this safe ? Doing that you give a static immutable reference, but you can modify its contents with `write`. Am I missing something ?

Reply:
I didn't think it would be writeable, but if it is, then it is indeed a problem. Got a suggestion on an alternative?

Reply:
You can't write through the reference, but you can use `FlashWriter::write` to modify the same memory that you're holding a reference to.
Hmm, maybe you can return a proxy type to prevent someone to call write and then use the ref. Maybe something like:
```rust
struct FlashReader<'a> {
    _flash: &'a FlashWriter,
    data: &'a [u8],
}

impl<'a> FlashReader<'a> {
    pub fn data(&self) -> &[u8] {
        self.data
    }
}
```
Then, you return that on the `FlashWriter::read` method.

And other things that I missed, I think `read` should only ask for `&self` instead of `&mut self` and also, have you tested if we can read a byte of a time instead of two ? The `Read Operation` part of the manual doesn't mention anything, so I think we can, but I didn't test it or read the entire manual.